partner_firstname/models.py

```diff
@@ -39,11 +40,13 @@ class ResPartner(models.Model):
         store=True)
 
     @api.one
-    @api.depends("firstname", "lastname")
+    @api.depends("firstname", "lastname", "company_id")
     def _compute_name(self):
         """Write the 'name' field according to splitted data."""
-        self.name = u" ".join((p for p in (self.lastname,
-                                           self.firstname) if p))
+        names = (self.lastname, self.firstname)
+        if self.company_id.names_order == 'first_last':
```

Comment:
Seems like you are merging the onchange fix with the new feature of choose names order. IMHO splitting that in 2 PR will make it easier to review.